### PR TITLE
fix orphaned pod flexvolume  can not be cleaned up

### PR DESCRIPTION
**Pull request by xiaoanyunfei**
**What type of PR is this?**

 /kind bug
/kind cleanup

**What this PR does / why we need it**:

The main reason is :

flexvolume `GetPath`  returns `hulk~lvm`, kubelet create mount path based on `GetPath`  .

but flexvolume is registered as `flexvolume-hulk~lvm`

kubelet can't find flexvolume plugin by  orphaned pod volume path like `/var/lib/kubelet/pods/bb07fd8d-b737-11e9-a203-0022ac9722f5/volumes/hulk~lvm`

flexVolumePluginNamePrefix must be blank now, otherwise kubelet will create a new volume for the existing pod

**Which issue(s) this PR fixes**:

Fixes #
https://github.com/kubernetes/kubernetes/issues/80972
**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```
/sig storage

### CODE REVIEW

File: pkg/kubelet/volumemanager/cache/actual_state_of_world.go [hunk 1]
```diff
@@ -498,26 +498,25 @@ func (asw *actualStateOfWorld) MarkRemountRequired(
...
-				// Log and continue processing
-				klog.Errorf(
-					"MarkRemountRequired failed to FindPluginBySpec for pod %q (podUid %q) volume: %q (volSpecName: %q)",
-					podObj.podName,
-					podObj.podUID,
-					volumeObj.volumeName,
-					podObj.volumeSpec.Name())
-				continue
-			}
+			if mountedPodName == podName {
```
Comment (tedyu) [inline pkg/kubelet/volumemanager/cache/actual_state_of_world.go]: Is this change necessary ?
It seems only indentation changed.

Comment (xiaoanyunfei) [reply pkg/kubelet/volumemanager/cache/actual_state_of_world.go]: en, I will remove it

**Comment by wgliang** [comment]
/ping @sig-node-reviewers
This is indeed a problem in our cluster. PTAL

**Comment by xiaoanyunfei** [comment]
I don't think pull-kubernetes-conformance-kind-ipv6 has anything to do with this pr

/assign @jsafrane 
please help to review it

File: pkg/volume/flexvolume/plugin-defaults.go [hunk 2]
```diff
@@ -28,7 +32,20 @@ func logPrefix(plugin *flexVolumePlugin) string {
...
+	hashutil.DeepHashObject(hash, path)
+	return uint64(hash.Sum32())
+}
+
 func (plugin *pluginDefaults) GetVolumeName(spec *volume.Spec) (string, error) {
 	klog.V(4).Infof(logPrefix((*flexVolumePlugin)(plugin)), "using default GetVolumeName for volume ", spec.Name())
+	// GetVolumeName should return an unique volume name as possible
+	if spec.Volume != nil && spec.Volume.FlexVolume != nil {
+		options := spec.Volume.FlexVolume.Options
+		if options != nil && options["mountpath"] != "" {
```
Comment (jsafrane) [inline pkg/volume/flexvolume/plugin-defaults.go]: This expects that the flex driver uses `mountpath` option. This may work for your driver, but different flex drivers use different options. We should not hardcode any options here.

The best would be if [GetVolumeName](https://github.com/kubernetes/kubernetes/blob/11d83dfdc17f0a80a7da73accfd060084cfee0a1/pkg/volume/flexvolume/plugin.go#L125) used the value provided by the driver, but it was removed in https://github.com/kubernetes/kubernetes/pull/47400. See also https://github.com/kubernetes/kubernetes/issues/44737, suggested json metadata file was never implemented.

@chakri-nelluri, do you have any other suggestion?
cc @gnufied

Comment (xiaoanyunfei) [reply pkg/volume/flexvolume/plugin-defaults.go]: I passed `mountpath` as argument of `volume.Spec` in `plugin.ConstructVolumeSpec`
I think it works for all flex driver,  although it doesn't look elegant enough.

There is only one place to call `plugin. GetVolumeName` once  when kubelet starts
The volume name only works when the pod is `orphan pod`  which will be cleaned up in `reconciler.reconcile` loop.

kubelet/volumemanager/reconciler/reconciler.go  `syncStates`  =>  `reconstructVolume` = > volume/util/operationexecutor/operation_executor.go `ReconstructVolumeOperation` => `plugin.ConstructVolumeSpec`

Comment (zhan849) [reply pkg/volume/flexvolume/plugin-defaults.go]: from code "unique volume name" is currently being generated anyway, and there is no semantic / convention that states `GetName()` should return something unique.

given how it's implemented now:
```
	var uniqueVolumeName v1.UniqueVolumeName
	if attachablePlugin != nil || deviceMountablePlugin != nil {
		uniqueVolumeName, err = util.GetUniqueVolumeNameFromSpec(plugin, volumeSpec)
		if err != nil {
			return nil, err
		}
	} else {
		uniqueVolumeName = util.GetUniqueVolumeNameFromSpecWithPod(volume.podName, plugin, volumeSpec)
	}
```

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/reconciler/reconciler.go#L490

shall we just always use `GetUniqueVolumeNameFromSpecWithPod` to generate unique name for volume reconciliation? we can include pod namespace as well so it's guaranteed that volume name in the map is unique

File: pkg/volume/flexvolume/plugin.go [hunk 3]
```diff
@@ -34,8 +34,9 @@ import (
 )
 
 const (
-	flexVolumePluginName       = "kubernetes.io/flexvolume"
-	flexVolumePluginNamePrefix = "flexvolume-"
+	flexVolumePluginName = "kubernetes.io/flexvolume"
+	// for historical reasons, keep blank to be compatible with past versions
+	flexVolumePluginNamePrefix = ""
```
Comment (zhan849) [inline pkg/volume/flexvolume/plugin.go]: shall we just simply remove this prefix? it's not exported and not included in volume path anyway.

**Comment by gaorong** [comment]
FYI: https://github.com/kubernetes/kubernetes/pull/71402

File: pkg/volume/flexvolume/plugin-defaults.go (hunk 2 above)
Comment (xiaoanyunfei) [reply pkg/volume/flexvolume/plugin-defaults.go]: It will affect all plugins not only flex volume

Comment (chakri-nelluri) [reply pkg/volume/flexvolume/plugin-defaults.go]: Options is not available in most of the code paths or when the mount is stale that is the reason GetVolumeName was disabled in the first place.
GetVolumeName should return unique name in the cluster except for inline volumes.
Are we seeing any issues after removing the prefix?

**Comment by xiaoanyunfei** [comment]
I gave up to change `GetVolumeName`,  maybe this new solution is more acceptable.

In order to be compatible with previous versions, `flexVolumePluginNamePrefix`  must be null, otherwise kubelet will create a new volume for the existing pod.

File: pkg/kubelet/volumemanager/reconciler/reconciler.go [hunk 4]
```diff
@@ -604,7 +609,7 @@ func getDeviceMountPath(volume *reconstructedVolume) (string, error) {
 	return deviceMountPath, nil
 }
 
-func (rc *reconciler) updateStates(volumesNeedUpdate map[v1.UniqueVolumeName]*reconstructedVolume) error {
+func (rc *reconciler) updateStates(volumesNeedUpdate []*reconstructedVolume) error {
```
Comment (chakri-nelluri) [inline pkg/kubelet/volumemanager/reconciler/reconciler.go]: This change should make it similar to any other plugin and I have not seen any issues with it so far with multiple vendor plugins.
Do we need the volumemanager changes?

Comment (xiaoanyunfei) [reply pkg/kubelet/volumemanager/reconciler/reconciler.go]: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/reconciler/reconciler.go#L414
`volumesNeedUpdate[reconstructedVolume.volumeName] = reconstructedVolume`

> GetVolumeName should return unique name in the cluster except for inline volumes.

The default Flexvolme `GetVolumeName` return `inline volume name`, and the pods on the node can use the same `inline volume name`.
When multiple volumes with same  `inline volume name` need to be cleaned up, only one can succeed if using `map volumesNeedUpdate`

`syncStates` is called once during the kubelet life cycle.

**Comment by xiaoanyunfei** [comment]
@chakri-nelluri  @jsafrane 
the related issue continues to appear recently, please have a look on this pr

File: pkg/kubelet/volumemanager/reconciler/reconciler.go (hunk 4 above)
Comment (chakri-nelluri) [reply pkg/kubelet/volumemanager/reconciler/reconciler.go]: @xiaoanyunfei This is a known limitation with inline & flex volumes. Inline volumes are not recommended for flex volumes. There is not enough information to reconstruct plugin information, unless we do Kubelet checkpointing.

Comment (xiaoanyunfei) [reply pkg/kubelet/volumemanager/reconciler/reconciler.go]: @chakri-nelluri 
The changes in `reconciler.go` don't affect other volume plugin, 
I just change the parameters format of `updateStates` and `updateDevicePath`.

 `This is a known limitation with inline & flex volumes. Inline volumes are not recommended for flex volumes.`

I don't get your point. Every flexvolume has an `inline volume name`, is that right?

So multiple flexvolumes in different pods can have the same `inline volume name`

If we have ten pods with flexVolumes which has same `inline volume name`(reconstructedVolume.volumeName in reconciler.syncStates)

 such as
>  name: hulklvm
     driver: hulk/lvm
      fsType: ext4
      options:
        size: 32Gi
        volumegroup: volumevg

when the pods become orphan pod, only one pod(in map `reconciler.syncStates volumesNeedUpdate`) can be clean up in the previous ` reconciler.go`

Comment (xiaoanyunfei) [reply pkg/kubelet/volumemanager/reconciler/reconciler.go]: @chakri-nelluri  

> Flexvolume inline volume names have to be cluster-wide unique if they are used. That is a known limitation

But k8s code don't limit this.  I think k8s should prepare for the worst.

Comment (xiaoanyunfei) [reply pkg/kubelet/volumemanager/reconciler/reconciler.go]: I will separate this file into another pr

**Comment by rphillips** [comment]
could you rebase to fix the conflicting files?

**Comment by rphillips** [comment]
cc @kubernetes/sig-storage-pr-reviews

**Review by thockin — COMMENTED** [review]
This seems OK, but @chakri-nelluri why was this prefix in there in the first place?

**Comment by thockin** [comment]
ping @chakri-nelluri ?

**Comment by chakri-nelluri** [comment]
@thockin Sorry Tim, didn't see the comment earlier.
This was added by @verult for tracking flex volume drivers from regular drivers. That caused other issues.

**Comment by thockin** [comment]
I have no real way to test this, so I will have to trust the review that has already been done :)

/approve

**Comment by wgliang** [comment]
A long journey, finally merged.
